Add quantized CELU operator by adding additional parameters to quantized ELU

durumu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39199 Add quantized CELU operator by adding additional parameters to quantized ELU**

Differential Revision: [D21771202](https://our.internmc.facebook.com/intern/diff/D21771202)

durumu committed:
Dummy operator building correctly

[ghstack-poisoned]

durumu committed:
Update on "Dummy operator building correctly"


Differential Revision: [D21771202](https://our.internmc.facebook.com/intern/diff/D21771202)

[ghstack-poisoned]

vkuzo commented:
there are some failing tests (like https://app.circleci.com/pipelines/github/pytorch/pytorch/174752/workflows/fa9a3d4b-602d-45b3-955f-8c32409c2eb5/jobs/5611015/steps), we should figure out what's going on with them before landing.  The failure looks odd to me, do we have a local repro?

vkuzo commented on torch/nn/quantized/functional.py:
```diff
@@ -478,9 +499,9 @@ def elu(input, alpha=1., inplace=False, scale=None, zero_point=None):
         return output
     elif inplace:
-        return torch._C._nn.elu_(input, alpha)
+        return torch._C._nn.elu_(input, alpha, 1., 1.)
```
these have a default of 1 in `native_functions.yaml`, would it work to revert this and next line?

vkuzo commented on aten/src/ATen/native/quantized/cpu/qelu.cpp:
```diff
@@ -29,8 +29,18 @@ Tensor quantized_elu(
 }
 
+Tensor& quantized_celu_(Tensor& self, Scalar alpha) {
+  double inv_alpha = 1. / alpha.to<double>();
```
can we verify that division by zero if alpha == 0 gives a good error message?

vkuzo reviewed (commented):
thanks for working on this!  in general looks great, we should figure out why the derivative tests are failing before landing

supriyar commented on torch/nn/quantized/functional.py:
```diff
@@ -362,6 +362,27 @@ def max_pool2d(input, kernel_size, stride=None, padding=0, dilation=1,
+    # type: (Tensor, Optional[float], Optional[bool]) -> Tensor
+    r"""celu(input, inplace=False) -> Tensor
+
+    Applies the quantized continuously differentiable ELU function element-wise.
```
ELU -> CELU

durumu replied:
CELU stands for continuously differentiable ELU -- maybe I should capitalize the C to make it clearer this is what I mean?

supriyar commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2601,8 +2601,16 @@
 - func: celu(Tensor self, Scalar alpha=1.0) -> Tensor
   use_c10_dispatcher: full
+  dispatch:
+    CPU: celu
+    CUDA: celu
```
Let's try removing these two lines for `celu` and `celu_` below

durumu committed:
Update on "Add quantized CELU operator by adding additional parameters to quantized ELU"


Differential Revision: [D21771202](https://our.internmc.facebook.com/intern/diff/D21771202)

[ghstack-poisoned]

durumu committed:
Update on "Add quantized CELU operator by adding additional parameters to quantized ELU"


Differential Revision: [D21771202](https://our.internmc.facebook.com/intern/diff/D21771202)

[ghstack-poisoned]

supriyar commented:
Changes look good! @vkuzo could you take a look as well?

vkuzo commented:
looks good, can we verify that `alpha == 0` is handled well with a good error message?

durumu committed:
Update on "Add quantized CELU operator by adding additional parameters to quantized ELU"


Differential Revision: [D21771202](https://our.internmc.facebook.com/intern/diff/D21771202)

[ghstack-poisoned]

durumu committed:
Update on "Add quantized CELU operator by adding additional parameters to quantized ELU"


Differential Revision: [D21771202](https://our.internmc.facebook.com/intern/diff/D21771202)

[ghstack-poisoned]

durumu commented:
@vkuzo I added a new error message that's more verbose - "ZeroDivisionError: alpha cannot be 0 in CELU".

vkuzo reviewed (approved):
looks great! thank you for working on this!

durumu committed:
Update on "Add quantized CELU operator by adding additional parameters to quantized ELU"


Differential Revision: [D21771202](https://our.internmc.facebook.com/intern/diff/D21771202)

[ghstack-poisoned]

ailzhang commented:
cc: @dlibenzi @JackCaoG  This PR might require pytorch/xla lowering `celu` .

dlibenzi commented:
> cc: @dlibenzi @JackCaoG This PR might require pytorch/xla lowering `celu` .

We chatted about this time ago as general discussion.
Would be nice if pytorch would leave default implementations of certain operations (especially the ones which already have implementations on the pytroch side, in terms of existing operators), so that not every backend is forced to implement them, if they are OK with the default implementation.

ailzhang commented:
> > cc: @dlibenzi @JackCaoG This PR might require pytorch/xla lowering `celu` .
> 
> We chatted about this time ago as general discussion.
> Would be nice if pytorch would leave default implementations of certain operations (especially the ones which already have implementations on the pytroch side, in terms of existing operators), so that not every backend is forced to implement them, if they are OK with the default implementation.

cc: @ezyang that we agreed this is the long term direction we want to achieve. But currently there's no way to allow this as ops have to go either device dispatch or typedefault.
I'll sync with @ezyang to see what's a good plan to get this done. 
But before that feature is available, pytorch/xla will have to move with pytorch master branch to lower it :(

dlibenzi commented:
> > > cc: @dlibenzi @JackCaoG This PR might require pytorch/xla lowering `celu` .
> > 
> > 
> > We chatted about this time ago as general discussion.
> > Would be nice if pytorch would leave default implementations of certain operations (especially the ones which already have implementations on the pytroch side, in terms of existing operators), so that not every backend is forced to implement them, if they are OK with the default implementation.
> 
> cc: @ezyang that we agreed this is the long term direction we want to achieve. But currently there's no way to allow this as ops have to go either device dispatch or typedefault.
> I'll sync with @ezyang to see what's a good plan to get this done.
> But before that feature is available, pytorch/xla will have to move with pytorch master branch to lower it :(

I guess one more entry in `aten_tensor_ops.cpp/h` 😞

ailzhang reviewed (approved):
I have a companion PR for this in pytorch/xla. Please feel free to rebase and land when it's ready.

durumu committed:
Update on "Add quantized CELU operator by adding additional parameters to quantized ELU"


Differential Revision: [D21771202](https://our.internmc.facebook.com/intern/diff/D21771202)

[ghstack-poisoned]

durumu committed:
Update on "Add quantized CELU operator by adding additional parameters to quantized ELU"


Differential Revision: [D21771202](https://our.internmc.facebook.com/intern/diff/D21771202)

[ghstack-poisoned]

ailzhang commented:
Btw a change of tensorpipe submodule is included in this PR which I think is not related.

supriyar commented:
> Btw a change of tensorpipe submodule is included in this PR which I think is not related.

@durumu could you remove the tensorpipe submodule update in the PR?

durumu committed:
Update on "Add quantized CELU operator by adding additional parameters to quantized ELU"


Differential Revision: [D21771202](https://our.internmc.facebook.com/intern/diff/D21771202)

[ghstack-poisoned]

durumu committed:
Update on "Add quantized CELU operator by adding additional parameters to quantized ELU"


Differential Revision: [D21771202](https://our.internmc.facebook.com/intern/diff/D21771202)

[ghstack-poisoned]

supriyar commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -793,7 +793,17 @@ void qtanh_kernel(const Tensor& qx, Tensor& qy) {
+void qelu_kernel(
+    const Tensor& qx,
+    Scalar alpha,
+    Scalar scale,
+    Scalar input_scale,
```
Can you try giving default values to these arguments.

durumu committed:
Update on "Add quantized CELU operator by adding additional parameters to quantized ELU"


Differential Revision: [D21771202](https://our.internmc.facebook.com/intern/diff/D21771202)

[ghstack-poisoned]

durumu committed:
Update on "Add quantized CELU operator by adding additional parameters to quantized ELU"


Differential Revision: [D21771202](https://our.internmc.facebook.com/intern/diff/D21771202)

[ghstack-poisoned]

durumu committed:
Update on "Add quantized CELU operator by adding additional parameters to quantized ELU"


Differential Revision: [D21771202](https://our.internmc.facebook.com/intern/diff/D21771202)

[ghstack-poisoned]

durumu committed:
Update on "Add quantized CELU operator by adding additional parameters to quantized ELU"


Differential Revision: [D21771202](https://our.internmc.facebook.com/intern/diff/D21771202)

[ghstack-poisoned]

durumu committed:
Update on "Add quantized CELU operator by adding additional parameters to quantized ELU"


Differential Revision: [D21771202](https://our.internmc.facebook.com/intern/diff/D21771202)

[ghstack-poisoned]

durumu committed:
Update on "Add quantized CELU operator by adding additional parameters to quantized ELU"


Differential Revision: [D21771202](https://our.internmc.facebook.com/intern/diff/D21771202)

[ghstack-poisoned]

supriyar commented on test/quantization/test_quantized_op.py:
```diff
@@ -325,6 +325,33 @@ def test_qelu(self, X, alpha):
+                                           dtype=torch_type)
+
+        # test regular
+        qY = torch.nn.quantized.functional.celu(qX, output_scale, output_zero_point, alpha=alpha)
```
I believe you can call `torch.ops.quantized.celu` here instead of functional

durumu committed:
Update on "Add quantized CELU operator by adding additional parameters to quantized ELU"


Differential Revision: [D21771202](https://our.internmc.facebook.com/intern/diff/D21771202)

[ghstack-poisoned]

durumu committed:
Update on "Add quantized CELU operator by adding additional parameters to quantized ELU"


Differential Revision: [D21771202](https://our.internmc.facebook.com/intern/diff/D21771202)

[ghstack-poisoned]

durumu committed:
Update on "Add quantized CELU operator by adding additional parameters to quantized ELU"


Differential Revision: [D21771202](https://our.internmc.facebook.com/intern/diff/D21771202)

[ghstack-poisoned]

durumu committed:
Update on "Add quantized CELU operator by adding additional parameters to quantized ELU"


Differential Revision: [D21771202](https://our.internmc.facebook.com/intern/diff/D21771202)

[ghstack-poisoned]

durumu committed:
Update on "Add quantized CELU operator by adding additional parameters to quantized ELU"


Differential Revision: [D21771202](https://our.internmc.facebook.com/intern/diff/D21771202)

[ghstack-poisoned]
